### PR TITLE
feat(client): add model diff for client

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description
add cmd: swcli model diff
image

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [x] add unit test
- [x] add necessary doc

### CODE REVIEW

File: client/starwhale/core/model/model.py [hunk 1]
```diff
@@ -374,6 +380,78 @@ def _get_module(cls, _model_config: ModelConfig) -> str:
...
+    def _to_file_desc(
+        self, files: t.List[t.Dict[str, t.Any]], parent_path: Path
+    ) -> t.Dict[str, FileDesc]:
+        return {
+            _f["path"]: FileDesc(
+                path=parent_path / _f["path"],
+                name=_f["name"] if "name" in _f else os.path.basename(_f["path"]),
+                size=_f["size"]
+                if "size" in _f
+                else (parent_path / _f["path"]).stat().st_size,
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/core/model/model.py]: yes, done

**Review by jialeicui — DISMISSED** [review]

File: client/starwhale/core/model/model.py [hunk 2]
```diff
@@ -389,6 +395,76 @@ def _get_module(cls, _model_config: ModelConfig) -> str:
...
+        if compare_uri.instance_type != InstanceType.STANDALONE:
+            raise NoSupportError(
+                f"only support standalone uri, but compare_uri({compare_uri}) is for cloud instance"
+            )
+        if self.uri.object.name != compare_uri.object.name:
+            raise NoSupportError(
+                f"only support two versions diff in one model, base model:{self.uri}, compare model:{compare_uri}"
+            )
+        _compare_model = StandaloneModel(compare_uri)
+        print(f"bb:{self.store.manifest['resources']}")
```
Comment (tianweidut) [inline client/starwhale/core/model/model.py:433]: use `console.print` ?

Comment (tianweidut) [reply client/starwhale/core/model/model.py:433]: 'bb' is  typo?

File: client/starwhale/core/model/model.py [hunk 3]
```diff
@@ -389,6 +395,76 @@ def _get_module(cls, _model_config: ModelConfig) -> str:
...
+            files=_compare_model.store.manifest["resources"],
+            parent_path=_compare_model.store.snapshot_workdir,
+        )
+        all_paths = set(
+            [p for m in (base_file_maps, compare_file_maps) for p, _ in m.items()]
+        )
+
+        for _p in all_paths:
+            if _p in base_file_maps and _p in compare_file_maps:
+                # updated or unchanged
```
Comment (tianweidut) [inline client/starwhale/core/model/model.py:448]: remove 448, and 454,459 lines comment?

File: client/starwhale/core/model/view.py [hunk 4]
```diff
@@ -33,6 +46,59 @@ def recover(self, force: bool = False) -> t.Tuple[bool, str]:
     def info(self, fullname: bool = False) -> None:
         self._print_info(self.model.info(), fullname=fullname)
 
+    def diff(self, compare_uri: URI, show_details: bool) -> None:
```
Comment (tianweidut) [inline client/starwhale/core/model/view.py:49]: add `@BaseTermView._only_standalone`  decorator for diff function?

File: client/starwhale/utils/fs.py [hunk 5]
```diff
@@ -121,6 +122,39 @@ def blake2b_file(fpath: t.Union[str, Path]) -> str:
     return _hash.hexdigest()
 
 
+def cmp_file_content(
+    base_path: t.Union[str, Path], cmp_path: t.Union[str, Path]
+) -> t.List[str]:
+    base_path = Path(base_path)
+    cmp_path = Path(cmp_path)
+    res = []
+    with base_path.open("r") as base:
```
Comment (tianweidut) [inline client/starwhale/utils/fs.py:131]:
```
    with base_path.open("r") as base:
        with cmp_path.open("r") as cmp:
```
-->
```
    with base_path.open("r") as base, cmp_path.open("r") as cmp:
```
?

File: client/starwhale/utils/fs.py [hunk 6]
```diff
@@ -121,6 +122,39 @@ def blake2b_file(fpath: t.Union[str, Path]) -> str:
...
+        with cmp_path.open("r") as cmp:
+            diff = difflib.unified_diff(
+                base.readlines(),
+                cmp.readlines(),
+                fromfile=os.path.basename(base_path),
```
Comment (tianweidut) [inline client/starwhale/utils/fs.py:136]: `os.path.basename(base_path)` -> `base_path.name` ?
`base_path` was already a Pathlib object

File: client/starwhale/utils/fs.py [hunk 7]
```diff
@@ -121,6 +122,39 @@ def blake2b_file(fpath: t.Union[str, Path]) -> str:
...
+                tofile=os.path.basename(cmp_path),
+                n=0,
+            )
+            for line in diff:
+
+                def skip(
+                    content: str, overview: bool = True, unchanged: bool = True
```
Comment (tianweidut) [inline client/starwhale/utils/fs.py:143]: skip function is a inner function, and `overview`, 'unchanged' arguments are not used, so we remove the useless code?

File: client/starwhale/core/model/model.py (hunk 2 above)
Comment (goldenxinxing) [reply client/starwhale/core/model/model.py:418]: removed

File: client/starwhale/core/model/model.py [hunk 8]
```diff
@@ -389,6 +395,76 @@ def _get_module(cls, _model_config: ModelConfig) -> str:
...
+        print(f"bb:{self.store.manifest['resources']}")
+        base_file_maps = self._to_file_desc(
+            files=self.store.manifest["resources"],
+            parent_path=self.store.snapshot_workdir,
+        )
+        compare_file_maps = self._to_file_desc(
+            files=_compare_model.store.manifest["resources"],
+            parent_path=_compare_model.store.snapshot_workdir,
+        )
+        all_paths = set(
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/core/model/model.py]: fixed

File: client/starwhale/core/model/view.py (hunk 4 above)
Comment (goldenxinxing) [reply client/starwhale/core/model/view.py:50]: fixed

File: client/starwhale/utils/fs.py (hunk 5 above)
Comment (goldenxinxing) [reply client/starwhale/utils/fs.py]: fixed

File: client/starwhale/utils/fs.py (hunk 6 above)
Comment (goldenxinxing) [reply client/starwhale/utils/fs.py]: fixed

File: client/starwhale/utils/fs.py (hunk 7 above)
Comment (goldenxinxing) [reply client/starwhale/utils/fs.py]: removed

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**